File: sdk/java/src/main/java/jp/kengos/shojiku/Request.java
package jp.kengos.shojiku;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The one JSON envelope every document operation crosses with.
 *
 * <p>Both entrances build it: sources resolved from a template NAME and sources the application
 * handed over as TEXT produce the same request, because the C surface has one request schema — and
 * that schema rejects unknown keys, so a key the engine may legitimately not receive is dropped
 * rather than sent as null.
 */
final class Request {

  private final Sources sources;
  private final Object params;
  private final String lang;
  private final List<String> fontDirs;
  private final List<String> localeDirs;

  Request(
      Sources sources, Object params, String lang, List<String> fontDirs, List<String> localeDirs) {
    this.sources = sources;
    this.params = params;
    this.lang = lang;
    this.fontDirs = fontDirs == null ? List.of() : List.copyOf(fontDirs);
    this.localeDirs = localeDirs == null ? List.of() : List.copyOf(localeDirs);
  }

  Request(Sources sources, Object params) {
    this(sources, params, null, null, null);
  }

  /**
   * The serialized envelope as UTF-8 bytes.
   *
   * <p>Params that cannot be serialized are programmer misuse — the engine's surface is UTF-8 JSON
   * by contract, so there is nothing to render — but letting a raw writer error escape {@code
   * generate} would make callers catch a foreign class they never invited into their code.
   */
  byte[] encoded() {
    try {
      return Json.write(envelope()).getBytes(StandardCharsets.UTF_8);
    } catch (UsageException error) {
      throw error;
    } catch (RuntimeException error) {
      throw new UsageException(
          "params could not be serialized as UTF-8 JSON: " + error.getMessage(), error);
    }
  }

  private Map<String, Object> envelope() {
    Map<String, Object> envelope = new LinkedHashMap<>();
    envelope.put("template", sources.template());
    envelope.put("params", paramsSource());
    envelope.put("fontDirs", fontDirs);
    envelope.put("localeDirs", localeDirs);

    // Absent rather than null: the request schema rejects unknown and ill-typed
    // keys, so a key the engine may legitimately not receive is dropped instead
    // of sent empty.
    if (sources.definitions() != null) {
      envelope.put("definitions", sources.definitions());
    }
    if (sources.assetsDir() != null) {
      envelope.put("assetsDir", sources.assetsDir());
    }
    if (lang != null) {
      envelope.put("lang", lang);
    }
    return envelope;
  }

  /**
   * A String params is the caller's own source text, passed through VERBATIM.
   *
   * <p>The engine parses JSON or YAML (YAML is a superset), so re-encoding it here would only be a
   * chance to change it. Anything else is serialized as JSON.
   *
   * <p>There is deliberately no per-format method family — format dispatch is the engine's, and an
   * SDK that offered {@code generateYaml} would be claiming a distinction the engine does not make.
   */
  private String paramsSource() {
    return params instanceof String text ? text : Json.write(params);
  }
}

File: engine/capi/src/request.rs
//! The request envelope every document operation takes.
//!
//! One JSON object rather than a dozen C parameters, for two reasons: the ABI
//! stays frozen while the envelope grows a key, and an SDK builds a map
//! instead of threading optional strings through a signature. The sources
//! travel INSIDE it as text — this library never reads a template off disk,
//! so there is no path handling here to get wrong.
//!
//! `deny_unknown_fields` is the point of the schema: a misspelled key in an
//! SDK is a located error, not a silently ignored setting. `scale` and
//! `pageIndex` are read by `shojiku_preview` alone; the other operations
//! accept and ignore them, which is what keeps ONE envelope type across the
//! surface.
//!
//! Two caps here are host-level guards rather than engine behaviour, and both
//! mirror what the MCP host already does with the same inputs: the
//! allow/deny id lists are bounded because the policy rescans them per asset,
//! and `scale` is bounded because it multiplies into a raster allocation.

use crate::status::Failure;
use serde::Deserialize;
use shojiku_authoring::fs::{resolve_font_dirs, resolve_locale_dir};
use shojiku_image::{AssetMode, AssetPolicy};
use std::path::{Path, PathBuf};

/// Most item ids one allow/deny list may carry, as in the MCP host.
const MAX_ASSET_IDS: usize = 256;
/// Widest accepted preview scale, in output pixels per layout point. The CLI
/// default is 2.0 (≈144 dpi); this bounds the raster a caller can ask for.
const MAX_SCALE: f64 = 10.0;
/// Preview scale when the request does not say — the CLI's default.
const DEFAULT_SCALE: f64 = 2.0;

/// One operation's inputs.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct Request {
    /// The template source.
    pub(crate) template: String,
    /// The definitions source, when the caller has one.
    #[serde(default)]
    pub(crate) definitions: Option<String>,
    /// The params source. Optional for `validate`, required to lay out.
    #[serde(default)]
    pub(crate) params: Option<String>,
    /// Locale id override, beating the template's own `defaults.locale`.
    #[serde(default)]
    pub(crate) lang: Option<String>,
    /// Extra font-pack search directories, highest priority first.
    #[serde(default)]
    font_dirs: Vec<PathBuf>,
    /// Extra locale-pack search directories, highest priority first.
    #[serde(default)]
    locale_dirs: Vec<PathBuf>,
    /// Directory bundled assets resolve against. Without it bundled sources
    /// are disabled — an inline template has no directory of its own.
    #[serde(default)]
    assets_dir: Option<PathBuf>,
    /// `"open"` (default) or `"bundled-only"`, the CLI's spellings.
    #[serde(default)]
    asset_mode: Option<String>,
    /// Item ids that may receive dynamic content under `bundled-only`.
    #[serde(default)]
    allow_dynamic_image: Vec<String>,
    /// Item ids that must never receive dynamic content.
    #[serde(default)]
    deny_dynamic_image: Vec<String>,
    /// Preview only: output pixels per layout point.
    #[serde(default)]
    scale: Option<f64>,
    /// Preview only: render just this 0-based page (the WASM host's
    /// convention, not the CLI's 1-based `--page`).
    #[serde(default)]
    pub(crate) page_index: Option<usize>,
}

impl Request {
    /// Reads the envelope. A malformed or unknown key comes back as an
    /// invalid-request failure carrying serde's own message, which names the
    /// offending field and where it was.
    pub(crate) fn parse(text: &str) -> Result<Request, Failure> {
        let request: Request =
            serde_json::from_str(text).map_err(|err| Failure::InvalidRequest(err.to_string()))?;
        request.check_asset_ids()?;
        request.check_scale()?;
        Ok(request)
    }

    /// The params source, which laying out cannot do without.
    pub(crate) fn require_params(&self) -> Result<&str, Failure> {
        self.params
            .as_deref()
            .ok_or_else(|| Failure::InvalidRequest("`params` is required to lay out".into()))
    }

    /// Font-pack search dirs: the request's, then `$SHOJIKU_FONT_DIR`, then
    /// the default — the same precedence the CLI gives its flags.
    pub(crate) fn font_dirs(&self) -> Vec<PathBuf> {
        resolve_font_dirs(&self.font_dirs)
    }

    /// Locale-pack search dirs, same precedence.
    pub(crate) fn locale_dirs(&self) -> Vec<PathBuf> {
        resolve_locale_dir(&self.locale_dirs)
    }

    /// The asset policy for this call; caps stay at the engine defaults.
    pub(crate) fn asset_policy(&self) -> Result<AssetPolicy, Failure> {
        Ok(AssetPolicy {
            mode: self.mode()?,
            dynamic_allow: self.allow_dynamic_image.clone(),
            dynamic_deny: self.deny_dynamic_image.clone(),
            ..AssetPolicy::default()
        })
    }

    /// The directory bundled assets resolve against, if any.
    pub(crate) fn assets_root(&self) -> Option<&Path> {
        self.assets_dir.as_deref()
    }

    /// The preview scale, defaulted.
    pub(crate) fn scale(&self) -> f64 {
        self.scale.unwrap_or(DEFAULT_SCALE)
    }

    /// `assetMode` in the CLI's spelling, so both hosts stay one surface.
    fn mode(&self) -> Result<AssetMode, Failure> {
        match self.asset_mode.as_deref() {
            None | Some("open") => Ok(AssetMode::Open),
            Some("bundled-only") => Ok(AssetMode::BundledOnly),
            // The refusal names the accepted values and never the rejected
            // one: it is caller-controlled text.
            Some(_) => Err(Failure::InvalidRequest(
                "`assetMode` must be \"open\" or \"bundled-only\"".into(),
            )),
        }
    }

    /// Bounds the allow/deny lists.
    fn check_asset_ids(&self) -> Result<(), Failure> {
        for (key, ids) in [
            ("allowDynamicImage", &self.allow_dynamic_image),
            ("denyDynamicImage", &self.deny_dynamic_image),
        ] {
            if ids.len() > MAX_ASSET_IDS {
                return Err(Failure::InvalidRequest(format!(
                    "`{key}` has {} entries, over the {MAX_ASSET_IDS}-entry cap",
                    ids.len()
                )));
            }
        }
        Ok(())
    }

    /// Bounds the preview scale. A non-finite scale would reach the
    /// rasterizer's own arithmetic, and a huge one is an allocation request
    /// wearing a float.
    fn check_scale(&self) -> Result<(), Failure> {
        let Some(scale) = self.scale else {
            return Ok(());
        };
        if scale.is_finite() && scale > 0.0 && scale <= MAX_SCALE {
            return Ok(());
        }
        Err(Failure::InvalidRequest(format!(
            "`scale` must be finite and within 0 < scale <= {MAX_SCALE}"
        )))
    }
}

#[cfg(test)]
mod tests;
